gcs/src/main/scala/monix/connect/gcp/storage/components/GcsUploader.scala

```diff
@@ -46,44 +45,61 @@ private[storage] final class GcsUploader(
   override def createSubscriber(
     cb: Callback[Throwable, Unit],
     s: Scheduler): (Subscriber[Array[Byte]], AssignableCancelable) = {
-    val sub = new Subscriber[Array[Byte]] {
+    val sub = new Subscriber.Sync[Array[Byte]] {
       self =>
       override implicit def scheduler: Scheduler = s
 
       val writer: WriteChannel = storage.writer(blobInfo, options: _*)
       writer.setChunkSize(chunkSize)
+      private[this] var isDone = false
 
-      override def onNext(chunk: Array[Byte]): Future[Ack] = {
-        Task {
+      override def onNext(chunk: Array[Byte]): Ack = {
+        try {
+          if (chunk.isEmpty) {
+            if (!isDone) { onComplete() }
+            Ack.Stop
+          } else {
+            writer.write(ByteBuffer.wrap(chunk))
+            monix.execution.Ack.Continue
+          }
+        } catch {
+          case NonFatal(ex) => {
+            onError(ex)
+            Ack.Stop
+          }
+        }
+      }
+
+      override def onComplete(): Unit = {
+        if (!isDone) {
+          isDone = true
           try {
-            if (chunk.isEmpty) {
-              onComplete()
-              Ack.Stop
-            } else {
-              writer.write(ByteBuffer.wrap(chunk))
-              monix.execution.Ack.Continue
-            }
+            writer.close()
+            cb.onSuccess(())
           } catch {
-            case ex if NonFatal(ex) => {
+            case NonFatal(ex) => {
               onError(ex)
-              Ack.Stop
             }
           }
-        }.runToFuture
+        }
       }
 
       override def onError(ex: Throwable): Unit = {
-        writer.close()
-        cb.onError(ex)
-      }
+        try {
+          writer.close()
+          cb.onError(ex)
+        } catch {
+          case NonFatal(ex2) => {
+            onError(ex)
+            cb.onError(Platform.composeErrors(ex, ex2))
+          }
+        }
 
-      override def onComplete(): Unit = {
-        writer.close()
-        cb.onSuccess(())
       }
+
     }
 
-    (sub, AssignableCancelable.dummy)
+    (sub, AssignableCancelable.multi())
```

Comment:
Why not `dummy`?

Reply:
because cancelable would't do anything, and with multi the ability to cancel is there (as per my understanding)

Reply:
@Avasil what would you recommend? I have changed that to `single`, since multiple is usually used for cancelling composed cancelable, which is not the case I think

Reply:
It should be `dummy` because there's nothing to cancel in the `Consumer` so we don't need this allocation.

It will be discarded in `consumeWith` so it won't negatively impact other cancelables (in Operators):

```scala
  final def apply(source: Observable[In]): Task[R] =
    Task.create { (scheduler, cb) =>
      val (out, consumerSubscription) = createSubscriber(cb, scheduler)
      // Start consuming the stream
      val sourceSubscription = source.subscribe(out)
      // Assign the observable subscription to our assignable,
      // thus the subscriber can cancel its subscription
      consumerSubscription := sourceSubscription
      // We might not return the assignable returned by `createSubscriber`
      // because it might be a dummy
      if (consumerSubscription.isInstanceOf[Cancelable.IsDummy])
        sourceSubscription
      else
        consumerSubscription
    }
```

(note the last if)

Reply:
Oh I see, thanks for clarifying :)